File: python/src/runapi/suno/resources/generate_artwork.py
"""Suno generate_artwork resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource

from .. import _validators
from ..types import CompletedGenerateArtworkResponse, GenerateArtworkResponse


class GenerateArtwork(Resource):
    """Generate cover artwork for a prior task."""

    ENDPOINT = "/api/v1/suno/generate_artwork"

    RESPONSE_CLASS = GenerateArtworkResponse
    COMPLETED_RESPONSE_CLASS = CompletedGenerateArtworkResponse

    def run(self, **params: Any) -> Any:
        """Create cover artwork and poll until it completes.

        Args:
            **params: artwork parameters.

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create a artwork task and return immediately with an id.

        Args:
            **params: artwork parameters.

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of a artwork task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        _validators.validate_generate_artwork(params)

File: ruby/lib/runapi/suno/resources/generate_artwork.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      # Creates cover artwork for an existing music task.
      class GenerateArtwork
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/generate_artwork"
        RESPONSE_CLASS = Types::GenerateArtworkResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedGenerateArtworkResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          Validators.validate_generate_artwork!(params, self)
        end
      end
    end
  end
end
